Feature/postselect-mbqc-demo

Roll249 opened the pull request:
Before submitting, please check the following:

- Make sure you have tests for the new code and that test passes (run `nox`)
- If applicable, add a line to the [unreleased] part of CHANGELOG.md, following [keep-a-changelog](https://keepachangelog.com/en/1.0.0/).
- Format added code by `ruff`
  - See `CONTRIBUTING.md` for more details
- Make sure the checks (github actions) pass.

Then, please fill in below:

**Context (if applicable):**

**Description of the change:**

**Related issue:**

EarlMilktea commented:
@Roll249 Could you remove `graph_state.png` ?

Roll249 commented:
@EarlMilktea oke i have delete that

mgarnier59 commented:
Hi @Roll249 ,
Thanks for your contribution. I will start having a look at what you did. Can you please detail what you have done and how it relates to the proposed reference? 
Note that we will soon merge a postselection mechanism.

Thanks!

Roll249 commented:
This pull request introduces post-selection functionality for measurements in the `Graphix` quantum simulator, allowing users to specify desired measurement outcomes for certain nodes. The changes include updates to the simulator's API, backend logic, and an example script demonstrating the feature.

### New Feature: Post-Selection Support
* [`graphix/pattern.py`](diffhunk://#diff-5095250b3743dab6bc338800ae655423e7d8a83ca1563393627c51aad58f19d7L1281-R1281): Added a `postselect` parameter to the `simulate_pattern` method, enabling users to specify a dictionary of node indices and their desired measurement outcomes. The parameter is passed to the `PatternSimulator` for execution. [[1]](diffhunk://#diff-5095250b3743dab6bc338800ae655423e7d8a83ca1563393627c51aad58f19d7L1281-R1281) [[2]](diffhunk://#diff-5095250b3743dab6bc338800ae655423e7d8a83ca1563393627c51aad58f19d7R1291-R1293) [[3]](diffhunk://#diff-5095250b3743dab6bc338800ae655423e7d8a83ca1563393627c51aad58f19d7L1300-R1303)
* [`graphix/simulator.py`](diffhunk://#diff-db003bfd0b0b43ac74f98d6d07d1fd18cb849fd5d077630bf74cb8c67a9aa35eL95-R95): Updated the `PatternSimulator` class to handle the `postselect` parameter. Included logic in the `run` method to enforce the specified outcomes during measurement, even in noisy scenarios. [[1]](diffhunk://#diff-db003bfd0b0b43ac74f98d6d07d1fd18cb849fd5d077630bf74cb8c67a9aa35eL95-R95) [[2]](diffhunk://#diff-db003bfd0b0b43ac74f98d6d07d1fd18cb849fd5d077630bf74cb8c67a9aa35eR140) [[3]](diffhunk://#diff-db003bfd0b0b43ac74f98d6d07d1fd18cb849fd5d077630bf74cb8c67a9aa35eR177-R184) [[4]](diffhunk://#diff-db003bfd0b0b43ac74f98d6d07d1fd18cb849fd5d077630bf74cb8c67a9aa35eR206-R212)

### Backend Enhancements for Post-Selection
* [`graphix/sim/base_backend.py`](diffhunk://#diff-1973fd29513ba44186c615226f19f1dcbc3b1c00be45aeea850b78bf95985affL210-R227): Modified the `measure` method to accept a `forced_result` parameter, allowing the backend to enforce specific measurement outcomes.
* [`graphix/sim/tensornet.py`](diffhunk://#diff-50e92562476d257012a0b1307568e6088eff6a2b73ada6c22cb8be06d2526a16L151-R151): Updated the `measure` method to support `forced_result` for tensor network simulations, ensuring compatibility with the new post-selection feature. [[1]](diffhunk://#diff-50e92562476d257012a0b1307568e6088eff6a2b73ada6c22cb8be06d2526a16L151-R151) [[2]](diffhunk://#diff-50e92562476d257012a0b1307568e6088eff6a2b73ada6c22cb8be06d2526a16R163-R182)

### Example Script
* [`examples/postselected_measurement.py`](diffhunk://#diff-1061ad6642ba7f07b730403751541ace43199f5c5ec63fb94c47ad6f5f3d54b0R1-R48): Added a new example script demonstrating how to create a simple graph state, perform a measurement with post-selection, and visualize the resulting state. This serves as a prototype for simulating non-unitary gates.

Roll249 commented:
@EarlMilktea check it

EarlMilktea commented:
> @EarlMilktea check it

@Roll249 Please resolve CI errors.

EarlMilktea commented on graphix/pattern.py:
```diff
@@ -1278,7 +1278,7 @@ def space_list(self):
 
     def simulate_pattern(
-        self, backend: str = "statevector", input_state: BasicStates = BasicStates.PLUS, **kwargs
+        self, backend: str = "statevector", input_state: BasicStates = BasicStates.PLUS, postselect: dict = None, **kwargs
```
Please use `collections.abc.Mapping` instead.

EarlMilktea commented on graphix/simulator.py:
```diff
@@ -173,7 +174,15 @@ def run(self, input_state=BasicStates.PLUS) -> None:
                     self.backend.entangle_nodes(edge=cmd.nodes)
                 elif cmd.kind == CommandKind.M:
-                    self.__measure_method.measure(self.backend, cmd)
+                    # Post-selection logic
```
Please add extra tests that cover your contributions.

thierry-martinez commented:
We already implemented a more general approach for branch selection that we did not released yet, sorry: I opened the PR #300. Anyway, this is only remotely related to #278.
